pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,578 @@
+		newHostChains = append(newHostChains, hostChains...)
+		newServiceChains = append(newServiceChains, svcChain)
+
+		// write chain and sticky session rule
+		for _, hostChain := range hostChains {
+			// Create chain
+			if chain, ok := existingChains[utiliptables.Chain(hostChain)]; ok {
+				chainsLines.WriteString(fmt.Sprintf("%s\n", chain))
+			} else {
+				chainsLines.WriteString(makeChainLine(hostChain))
+			}
+
+			// Sticky session
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --rcheck --seconds %d --reap -j %s\n", svcChain, name.String(), hostChain, info.stickyMaxAgeMinutes*60, hostChain))
+			}
+		}
+
+		// write proxy/loadblanacing rules
+		n := len(hostChains)
+		for i, hostChain := range hostChains {
+			// Roughly round robin statistically if we have more than one host
+			if n > 1 {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m statistic --mode random --probability %f -j %s\n", svcChain, name.String(), 1.0/float64(n-i), hostChain))
+			} else {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j %s\n", svcChain, name.String(), hostChain))
+			}
+			// proxy
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --set -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), hostChain, protocol, hosts[i]))
+			} else {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), protocol, hosts[i]))
+			}
+		}
+
+		// proxy
+		rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), info.portal.ip.String(), protocol, protocol, info.portal.port, svcChain))
+
+		for _, publicIP := range info.deprecatedPublicIPs {
+			rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"deprecated-PublicIP portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), publicIP, protocol, protocol, info.portal.port, svcChain))
+		}
+
+		for _, ingress := range info.loadBalancerStatus.Ingress {
+			if ingress.IP != "" {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"load-balancer portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), ingress.IP, protocol, protocol, info.portal.port, svcChain))
+			}
+		}
+	}
+
+	// Delete chains no longer in use:
+	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
+	for _, chain := range newHostChains {
+		activeChains[chain] = true
+	}
+	for _, chain := range newServiceChains {
+		activeChains[chain] = true
+	}
+	for chain := range existingChains {
```

Comment:
Ah, that makes sense. It should throw an error in the logs in the rules
aren't supported.
Easy fix then, flush and delete. Will have in next push.

On Wed, Aug 12, 2015 at 1:30 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36827676>
> :
>
> > +		for _, ingress := range info.loadBalancerStatus.Ingress {
> > +			if ingress.IP != "" {
> > +				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"load-balancer portal for %s\" -d %s/32 -m state --state NEW -p %s -m %s --dport %d -j %s\n", iptablesServicesChain, name.String(), ingress.IP, protocol, protocol, info.portal.port, svcChain))
> > +			}
> > +		}
> > +	}
> > +
> > +	// Delete chains no longer in use:
> > +	activeChains := make(map[utiliptables.Chain]bool) // use a map as a set
> > +	for _, chain := range newHostChains {
> > +		activeChains[chain] = true
> > +	}
> > +	for _, chain := range newServiceChains {
> > +		activeChains[chain] = true
> > +	}
> > +	for chain := range existingChains {
>
> Oh, it DOES support -X, you just also have to -F those chains.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36827676>
> .
>